### PR TITLE
llama3.2 + cross attn test

### PR DESCRIPTION
**Pull request by maleksan85**
Commands:
server
```
 VLLM_NO_TUNED_GEMM=1 vllm serve /data/models/Llama-3.2-90B-Vision-Instruct --tensor_parallel_size 2 --enforce-eager --limit-mm-per-prompt "image=2" --max-num-seqs 32 --max_model_len 8192
```
client (from https://huggingface.co/nltpt/VLLM-llama3.2):
```
root@banff-cyxtera-s82-5:~/workspace/VLLM-llama3.2# python openai_vision_api_client.py
Chat completion output: The image depicts a serene lake scene with a wooden dock extending into the water, surrounded by lush greenery and majestic mountains in the background. The overall atmosphere of the image exudes tranquility and natural beauty, inviting the viewer to step into its peaceful world.
remove me: testing done, exitting...
```

### CODE REVIEW

File: tests/kernels/test_encoder_decoder_attn.py [hunk 1]
```diff
@@ -807,7 +808,7 @@ def test_encoder_only(
         assert_actual_matches_ideal(enc_test_params, enc_pckd_act_out)
 
 
-@pytest.mark.skipif(is_hip(), reason=STR_NOT_IMPL_ENC_DEC_ROCM_HIP)
+# @pytest.mark.skipif(is_hip(), reason=STR_NOT_IMPL_ENC_DEC_ROCM_HIP)
```
Comment (shajrawi) [inline tests/kernels/test_encoder_decoder_attn.py:811]: why is this commented out? either keep it or remove it?

**Review by shajrawi — COMMENTED** [review]
Great work!!
Few nits + one question.
Also, can you please make the linter happy? :)

File: vllm/model_executor/layers/linear.py [hunk 2]
```diff
@@ -131,7 +134,10 @@ def apply(self,
               layer: torch.nn.Module,
               x: torch.Tensor,
               bias: Optional[torch.Tensor] = None) -> torch.Tensor:
-        return tgemm.mm(x, layer.weight, bias)
+        if envs.VLLM_NO_TUNED_GEMM:
+            return F.linear(x, layer.weight, bias)
+        else:
+            return tgemm.mm(x, layer.weight, bias)
```
[the remark this answers is not in the document]
Comment (gshtras) [reply vllm/model_executor/layers/linear.py:140]: If certain 3.2 shapes have dimensions unsupported by hipb_mm, it's best to determine that based on the shape and not env globally

File: tests/kernels/test_encoder_decoder_attn.py (hunk 1 above)
Comment (maleksan85) [reply tests/kernels/test_encoder_decoder_attn.py:811]: removed, thanks

File: tests/kernels/utils.py [hunk 3]
```diff
@@ -524,8 +524,13 @@ def make_backend(backend_name: str) -> AttentionBackend:
     if backend_name == STR_XFORMERS_ATTN_VAL:
         # NOTE: xFormers backend cannot be imported for CPU and AMD GPUs.
         from vllm.attention.backends.xformers import XFormersBackend
-
         return XFormersBackend()
+    
+    if backend_name == STR_ROCM_FLASH_ATTN_VAL:        
+        from vllm.attention.backends.rocm_flash_attn import (  # noqa: F401
```
[the remark this answers is not in the document]
Comment (maleksan85) [reply tests/kernels/utils.py:530]: this is for linter to ignore unused import, like in here:
https://github.com/ROCm/vllm/blob/48c0cb4f2f4d8aa09c80a90faea72cd8d093a878/vllm/attention/backends/rocm_flash_attn.py#L346

File: vllm/attention/backends/rocm_flash_attn.py [hunk 4]
```diff
@@ -2,6 +2,7 @@
 from dataclasses import dataclass
 from typing import TYPE_CHECKING, Any, Dict, List, Optional, Tuple, Type
 
+from tests.kernels.utils import ref_masked_attention
```
[the remark this answers is not in the document]
Comment (maleksan85) [reply vllm/attention/backends/rocm_flash_attn.py]: removed, during implementation wanted to check correctness by running pytorch version of attention.

File: vllm/attention/backends/rocm_flash_attn.py [hunk 5]
```diff
@@ -377,7 +474,7 @@ def repeat_kv(self, x: torch.Tensor, n_rep: int) -> torch.Tensor:
                   None, :].expand(tokens, n_kv_heads, n_rep,
                                   head_dim).reshape(tokens, n_kv_heads * n_rep,
                                                     head_dim))
-
+
```
[the remark this answers is not in the document]
Comment (maleksan85) [reply vllm/attention/backends/rocm_flash_attn.py:484]: sure, emptied the line. But line is needed to separate new function definition.

File: vllm/model_executor/layers/linear.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (maleksan85) [reply vllm/model_executor/layers/linear.py:139]: it says:
```
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1747, in _call_impl
    return forward_call(*args, **kwargs)
  File "/root/workspace/vllm/vllm/model_executor/layers/linear.py", line 371, in forward
    output_parallel = self.quant_method.apply(self, input_, bias)
  File "/root/workspace/vllm/vllm/model_executor/layers/linear.py", line 139, in apply
    return tgemm.mm(x, layer.weight, bias)
  File "/root/workspace/vllm/vllm/model_executor/layers/tuned_gemm.py", line 87, in mm
    inp_view = inp.view(-1, inp.size(-1))
RuntimeError: view size is not compatible with input tensor's size and stride (at least one dimension spans across two contiguous subspaces). Use .reshape(...) instead.

```

[the remark this answers is not in the document]
Comment (maleksan85) [reply vllm/model_executor/layers/linear.py]: ok. replaced view to reshape. This might imply extra data move as reshape trying to make data contiguous.

**Review by shajrawi — APPROVED** [review]
Looks good - assuming performance does not regress due to reshape

File: vllm/model_executor/layers/linear.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (gshtras) [reply vllm/model_executor/layers/linear.py]: Copying the tensor is likely to be a huge perf impact. I think just fall back to F.linear if it isn't contiguous for now, we'll think of optimization in the future

[the remark this answers is not in the document]
Comment (shajrawi) [reply vllm/model_executor/layers/linear.py]: Yeah I was just talking about reshape on chat :D this got me nervous - Alex, let's go with Greg's suggestion here.

**Review by gshtras — APPROVED** [review]

**Pull request merged**